lib/compatibility.bzl

```diff
@@ -0,0 +1,64 @@
+"""TODO(trybka): Write module docstring."""
+
+load(":selects.bzl", "selects")
+
+INCOMPATIBLE_TARGET="@platforms//:incompatible"
+INCOMPATIBLE_SETTING="@platforms//:incompatible_setting"
+
+def _none_of_without_no_match(settings):
+    settings_name = "_or_".join([s.split(":")[1] for s in settings])
+    name = settings_name.replace(":", "-").replace("//", "").replace("/", "_")
+    compat_name = "incompatible_with_" + name
+    if not native.existing_rule(compat_name):
+        # We need to make (or re-use?) constraint values within a single instance to avoid a duplicate label.
+        native.constraint_value(name = compat_name, constraint_setting = INCOMPATIBLE_SETTING)
+    return selects.with_or({"//conditions:default": [], tuple(settings): [":" + compat_name]})
+
+def _any_of_without_no_match(settings):
+    settings_name = "_or_".join([s.split(":")[1] for s in settings])
+    name = settings_name.replace(":", "-").replace("//", "").replace("/", "_")
+    compat_name = "compatible_with_any_of_" + name
+    if not native.existing_rule(compat_name):
+        native.constraint_value(name = compat_name, constraint_setting = INCOMPATIBLE_SETTING)
+    return selects.with_or({tuple(settings): [], "//conditions:default": [":" + compat_name]})
+
+def _all_of_without_no_match(settings):
+    settings_name = "_and_".join([s.split(":")[1] for s in settings])
+    name = settings_name.replace(":", "-").replace("//", "").replace("/", "_")
+    if not native.existing_rule(name):
+        selects.config_setting_group(name = name, match_all = settings)
+    compat_name = "compatible_with_all_of_" + name
+    if not native.existing_rule(compat_name + name):
+        native.constraint_value(name = compat_name, constraint_setting = INCOMPATIBLE_SETTING)
+    return select({":" + name: [], "//conditions:default": [":" + compat_name]})
+
+verbose_compatibility = struct(
+    all_of = _all_of_without_no_match,
+    any_of = _any_of_without_no_match,
+    none_of = _none_of_without_no_match,
+)
+
+# Another alternative:
+# def compatibility(any_of = [], none_of = [], all_of = None):
+
+
+def _none_of(settings):
+    return selects.with_or({"//conditions:default": [], tuple(settings): [INCOMPATIBLE_TARGET]})
+
+def _any_of(settings):
+    no_match_error = "Didn't match any conditions." # TODO(trybka): concat setting names?
+    return selects.with_or({tuple(settings): [],}, no_match_error = no_match_error)
+
+def _all_of(settings):
+    settings_name = "_and_".join([s.split(":")[1] for s in settings])
+    name = settings_name.replace(":", "-").replace("//", "").replace("/", "_")
+    if not native.existing_rule(name):
+        selects.config_setting_group(name = name, match_all = settings)
```

Comment:
Yeah, this makes sense. I never tried to implement `all_of` before, but it makes sense that you can't do it with a single `select()` statement.

Reply:
It's cool to see composition of Starlark libs actually working.  I always worry the "non-native" part of `selects` means using `config_setting_group` as a library call could blow up in unexpected ways. Happy to see it all remains smooth.

Reply:
I think the only sharp/weird edge here is the fact that this has to be a named rule (which necessitated the `existing_rule` check).  If you have alternatives for how to make unique or unnamed targets here, that would improve things, I think.

Reply:
So... I had a thought. Would a simple implementation of `all_of` look like this?
```python
def _all_of(settings):
    return settings
```
The `target_compatible_with` attribute already implements AND logic by default.

Maybe as per your comment below, maybe don't provide `all_of` at all? I like it there for consistency, but maybe it's superfluous?

Reply:
That would only work if `settings` contained only `constraint_values`. In the case I want to support, raw `config_settings` don't work here. (Apparently anything that provides `ConstraintValueInfo` would work, but currently nothing else does, and that provider is behind some flag anyhow.)

Reply:
Ah dang. Yeah, I keep forgetting about that, apologies!